app/models/cherrypick_task.rb

```diff
@@ -33,23 +33,23 @@ def control_positions(batch_id, num_plate, total_wells, num_control_wells, wells
 
     raise StandardError, 'More controls than free wells' if num_control_wells > total_available_positions
 
-    quotient = batch_id
-    size_region = (total_available_positions / num_control_wells)
-    regions = (wells_to_leave_free..(total_wells - 1)).each_slice(size_region).to_a
+    available_positions = (wells_to_leave_free..(total_wells - 1)).to_a
+    # If num plate is equal to the available positions, the cycle is going to be repeated.
+    # To avoid it, every num_plate=available_positions we start a new cycle with a new seed.
+    seed = batch_id * ((num_plate / available_positions.length) + 1)
+    initial_positions = random_elements_from_list(available_positions, num_control_wells, seed)
```

Comment:
You could avoid the need to do this by moving `Random.new(seed)` out of `random_elements_from_list` and to the top of the `auto_add_control_plate` of `perform_pick`.
We then pass the PRG around in place of batch id, and don't even need to worry about passing in the plate number, the RNG will ensure we don't start cycling.
(Good call on the use Random.new and sample, was planning to do something similar when I glanced at your code review over the holiday)

Reply:
Hmm, thinking about it, this might even be a good chance to create a separate object to handle it... I'll mock one up

Reply:
```
class ControlLocator
  def initialize(seed, total_wells, number_controls, wells_to_leave_free: 0)
    @seed = Random.new(seed)
    @number_controls = number_controls
    @available_wells = (wells_to_leave_free...total_wells).to_a
  end

  # Returns the next set of control positions.
  # Subsequent calls will return a new set of positions 
  def control_positions!(number = number_controls)
    @available_wells.sample(number, random: @seed)
  end
end

# Usage
# def perform_pick(requests, robot, auto_add_control_plate, workflow_controller)
# ...
# control_assets = auto_add_control_plate&.wells&.joins(:samples) || 0
# control_locator = ControlLocator.new(batch.id, current_destination_plate.size, control_assets.count)
#
# Then each time you need them
# control_posns = control_locator.control_positions!

```

Reply:
Yes, that would make everything totally random for every plate, however that changes the actual behaviour so every plate in the same batch have a kind of predictable order in the controls positioning. The original story was:

https://github.com/sanger/sequencescape/issues/2783

There has been some changes in requirements but I will check with the SSRs what they think about this, so ask them between the 2 options:

- Option A: We generate random position for first plate, but next plates have a relation with this position by increasing 1, so they can use it to identify plate swaps, if that happens.
- Option B: We generate random positions for every batch and every single plate, independently of the position, so it has more variability

If they want to identify plate swaps, still in both options the controls have a predictable position, but in Option B could be easier to check.

Reply:
Mentioned in the slack, but I misunderstood exactly what was happening here. Hadn't realised that we only generate a new seed every x plates.